keps/sig-node/3008-cri-class-based-resources/README.md

```diff
@@ -0,0 +1,1550 @@
+may be set and they represent the Pod-level assignment of QoS-class resources,
+comparable to the PodClassResources message in PodSandboxConfig in the CRI API.
+
+```diff
+ type PodSpec struct {
+@@ -224,4 +224,8 @@ type PodSpec struct {
+     // Default to false.
+     // +optional
+     SetHostnameAsFQDN *bool `json:"setHostnameAsFQDN,omitempty" protobuf:"varint,35,opt,name=setHostnameAsFQDN"`
++    // Pod-level resources. Currently, requests and limits are not allowed
++    // to be specified for pods.
++    // +optional
++    Resources ResourceRequirements
+ }
+```
+
+There is already an ongoing effort to add [Pod level resource limits](#1592)
+that aims at adding a pod level `Resources` field in a similar fashion.
+
+In practice, the QoS-class resource information will be directly used in the CRI
+ContainerConfig (e.g.  CreateContainerRequest message). At this point, without
+resource discovery or access control kubelet does not do any validity checking
+of the values. Invalid class assignments will cause an error in the container
+runtime which causes the corresponding CRI RuntimeService request (e.g.
+RunPodSandbox or CreateContainer) to fail with an error.
+
+This phase would likely also wire QoS-class resources to
+[In-place pod vertical scaling](#1287), allowing updates of running containers.
+
+Input validation of classes very similar to labels is implemented: keys
+(`ClassResourceName`) and values must be non-empty, less than 64 characters
+long, must start and end with an alphanumeric character and may contain only
+alphanumeric characters, dashes, underscores or dots (`-`, `_` or `.`).
+Similar to labels, a namespace prefix (FQDN subdomain separated with a slash)
+in the key is allowed, similar to labels, e.g. `vendor/resource`.
+
+#### Update sandbox-level QoS-class resources
+
+Currently there is no endpoint in the CRI API to update the configuration of
+pod sandboxes. In contrast, container-level resources can be updated with the
+UpdateContainerResources API endpoint. In order to make container and pod
+(sandbox) level QoS-class resources symmetric we want to make it possible to
+update of pod-level resource assignments, too.
+
+This will likely required a new API endpoint in CRI:
+
+```diff
+@@ -38,6 +38,8 @@ service RuntimeService {
+     // RunPodSandbox creates and starts a pod-level sandbox. Runtimes must ensure
+     // the sandbox is in the ready state on success.
+     rpc RunPodSandbox(RunPodSandboxRequest) returns (RunPodSandboxResponse) {}
++    // UpdatePodSandboxConfig updates the configuration of an existing pod sandbox.
++    rpc UpdatePodSandboxConfig(UpdatePodSandboxConfigŔequest) returns (UpdatePodSandboxConfigŔesponse) {}
+```
+
+#### Resource status/capacity
+
+This KEP does not speak out anything about presenting the available resource
+types (or classes within) to the users.
+
+Some alternatives for presenting this information:
+
+1. Supplement `NodeStatus`
```

Comment:
1: should we include sig-scheduling in this conversation?
2: does it make sense to have the presentation of the information be in a later release (beta criteria, for instance)?

Reply:
> 1: should we include sig-scheduling in this conversation?

Definitely(!) But I think in the next implementation phase when we do something else than just cri-api changes.

> 2: does it make sense to have the presentation of the information be in a later release (beta criteria, for instance)?

Yeah, all of these (pod spec, node status etc) are **Future work**. I now tried to make this aspect more evident by mentioning *future step* in the beginning of each sub-section. The immediate changes (impl. phase 1) is just the cri-api plus annotations.

Reply:
resolved now as NodeStatus, PodStatus et al. are now defined and part of the first implementation phase